schema_compatibility.go

```diff
@@ -230,12 +232,192 @@ func (c *SchemaCompatibility) contains(a []string, s string) bool {
 	return false
 }
 
-func (c *SchemaCompatibility) getField(a []*Field, f *Field) (*Field, bool) {
+type getFieldOptions struct {
+	fieldAlias bool
+	elemAlias  bool
+}
+
+func (c *SchemaCompatibility) getField(a []*Field, f *Field, optFns ...func(*getFieldOptions)) (*Field, bool) {
+	opt := getFieldOptions{}
+	for _, fn := range optFns {
+		if fn == nil {
+			continue
+		}
+		fn(&opt)
+	}
 	for _, field := range a {
 		if field.Name() == f.Name() {
 			return field, true
 		}
+		if opt.fieldAlias {
+			for _, alias := range f.Aliases() {
+				if field.Name() == alias {
+					return field, true
+				}
+			}
+		}
+		if opt.elemAlias {
+			for _, alias := range field.Aliases() {
+				if f.Name() == alias {
+					return field, true
+				}
+			}
+		}
 	}
 
 	return nil, false
 }
+
+func isNative(typ Type) bool {
+	switch typ {
+	case Null, Boolean, Int, Long, Float, Double, Bytes, String:
+		return true
+	default:
+	}
+
+	return false
+}
+
+func isPromotable(typ Type) bool {
+	switch typ {
+	case Int, Long, Float, String, Bytes:
+		return true
+	default:
+	}
+
+	return false
+}
+
+func (c *SchemaCompatibility) Resolve(reader, writer Schema) (Schema, error) {
+	if reader.Type() == Ref {
+		reader = reader.(*RefSchema).Schema()
+	}
+	if writer.Type() == Ref {
+		writer = writer.(*RefSchema).Schema()
+	}
+
+	if err := c.compatible(reader, writer); err != nil {
```

Comment:
You call this recursively, even though it checks all sub schemas. I would suggest you split this function into 2, the starter that calls this, and the recursive (perhaps `resolve`).

Reply:
You're right! I noticed that it has a cache mechanism, so somehow, I tolerated the recursive call…